pkg/ruler/ruler.go

```diff
@@ -1320,3 +1333,40 @@ func returnByState(requestState string, alertState string) bool {
 func returnByHealth(requestHealth string, ruleHealth string) bool {
 	return requestHealth == "" || requestHealth == ruleHealth
 }
+
+func parseMatchersParam(matchers []string) ([][]*labels.Matcher, error) {
+	var matcherSets [][]*labels.Matcher
+	for _, s := range matchers {
+		matchers, err := parser.ParseMetricSelector(s)
+		if err != nil {
+			return nil, err
+		}
+		matcherSets = append(matcherSets, matchers)
+	}
+
+OUTER:
+	for _, ms := range matcherSets {
+		for _, lm := range ms {
+			if lm != nil && !lm.Matches("") {
+				continue OUTER
+			}
+		}
+		return nil, errors.New("match[] must contain at least one non-empty matcher")
+	}
+	return matcherSets, nil
+}
+
+func matches(matcherSets [][]*labels.Matcher, l labels.Labels) bool {
+	if len(matcherSets) == 0 {
+		return true
+	}
+
+	for _, matchers := range matcherSets {
+		for _, m := range matchers {
+			if v := l.Get(m.Name); m.Matches(v) {
+				return true
+			}
```

Comment:
This seems wrong. I understand that the rule labels need to match all matchers of `matchers` if the label name exists so it is an `AND`.
Only `matcherSets` is an `OR`.

https://github.com/prometheus/prometheus/pull/10194/files#diff-9008036c169df3fe4143c58b08426516bd3fd135f182cacc8b9326541fa55235R167 seems the correct implementation.

Reply:
Let's also add unit tests to cover this function

Reply:
good catch!